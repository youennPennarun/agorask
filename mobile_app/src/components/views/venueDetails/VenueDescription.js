import React, {PropTypes} from 'react';
import {View, Text, StyleSheet, Dimensions} from 'react-native';
import Icon from 'react-native-vector-icons/Entypo';

const {width} = Dimensions.get('window');

function getIcon(amenity) {
  if (amenity === 'pub') return 'drink';
  return 'globe';
}
function renderAddress(formattedAddress) {
  if (typeof formattedAddress === 'string') {
    return (<Text>{formattedAddress}</Text>);
  } else if (formattedAddress instanceof Array) {
    return (
      <View>{formattedAddress.map((line, key) => <Text key={key}>{line}</Text>)}</View>
    );
  }
  return <View />;
}

export default function VenueDescription(props) {
  const {name, address, categories = [], website, contact = {formattedPhone: ''}} = props.venue;
  const categoryName = (categories.length) ? categories[0].name : '';
  return (
    <View style={styles.block}>
      <View style={styles.titleContainer} >
        <Icon name={getIcon()}
          style={styles.icon}
          size={16}
          color='blue' />
        <Text style={styles.venueName}>{name}</Text>
      </View>
      {renderAddress(address.formatted)}
      <Text>{categoryName}</Text>
      <Text>{website}</Text>
      {(contact.formattedPhone) ? (
        <Text><Icon name={'mobile'}
          style={styles.icon}
          size={16}
          color='blue' />{contact.formattedPhone}
        </Text>) : null
      }
    </View>
  );
}

const styles = StyleSheet.create({
  block: {
    width,
    backgroundColor: 'white',
    marginTop: 5,
    paddingTop: 10,
    paddingLeft: 10,
    paddingRight: 10,
    paddingBottom: 10,
  },
  icon: {
    marginRight: 15,
  },
  titleContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  venueName: {
    fontSize: 20,
  },
});

VenueDescription.propTypes = {
  venue: PropTypes.object.isRequired,
};
