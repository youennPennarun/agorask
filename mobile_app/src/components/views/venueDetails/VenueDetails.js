import React, {Component, PropTypes} from 'react';
import {View, Image, Text, StyleSheet, Dimensions, ScrollView} from 'react-native';

import {connect} from 'react-redux';

import Icon from 'react-native-vector-icons/EvilIcons';
import update from 'immutability-helper';

import RoundButton from '../../commons/RoundButton';

import AddTask from './task/AddTask';

import { graphql } from 'react-apollo';
import gql from 'graphql-tag';


import VenueDescription from './VenueDescription';
import Tasks from './task/Tasks';


const {width, height} = Dimensions.get('window');

function renderFetchingState() {
  return <View />;
}

function renderError(error) {
  console.log(error);
  return (
    <View>
      <Text>Error: {error}</Text>
    </View>
  )
}

export class VenueDetails extends Component {
  state = {
    showAddTask: false,
  }

  addTask(task) {
    this.props.addTask(this.props.venue._id, task, this.props.token);
  }

  render() {
    const {venue, isFetching, error} = this.props;
    if (isFetching) return renderFetchingState();
    if (error) return renderError(error);
    let imageUri = 'http://www.eltis.org/sites/eltis/files/default_images/photo_default_4.png';
    if (venue._id) {
      imageUri = `http://192.168.0.10:3000/venues/${venue._id}/image`;
    }
    return (
      <View style={styles.container}>
        <ScrollView style={styles.scollView}>
          <Image style={styles.coverImage}
            resizeMode='cover'
            source={{uri: imageUri}} />
          <VenueDescription venue={venue} />
          <Tasks tasks={venue.tasks || []}
            goToTask={(id, task) => { this.props.navigator.taskDetails(id, task); }} />
        </ScrollView>
        <AddTask style={styles.addTask}
          visible={this.state.showAddTask}
          addTask={task => { this.addTask(task); }} />

        <RoundButton style={styles.addTaskBtn}
          size={50}
          initialState={{
            color: 'green',
            renderIcon: () => <Icon name='plus' size={42} color='white' />,
            callback: () => this.setState({showAddTask: true}),
          }}
          activatedState={{
            color: 'red',
            renderIcon: () => <Icon name='close' size={42} color='white' />,
            callback: () => this.setState({showAddTask: false}),
          }} />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#e9e9e9',
  },
  coverImage: {
    width: width,
    height: height * 0.3,
  },
  scollView: {
    height,
  },
  block: {
    backgroundColor: 'white',
    marginTop: 5,
    paddingTop: 10,
    paddingLeft: 10,
    paddingRight: 10,
    paddingBottom: 10,
  },
  titleContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  venueName: {
    fontSize: 20,
  },
  addTaskBtn: {
    width: 50,
    height: 50,
    borderRadius: 25,
    alignItems: 'center',
    justifyContent: 'center',
    position: 'absolute',
    bottom: 5,
    right: 5,
    elevation: 20,
  },
  addTaskBtnLabel: {
    fontSize: 22,
    color: 'white',
  },
  addTask: {
  }
});

VenueDetails.propTypes = {
  _id: PropTypes.string,
  sourceId: PropTypes.string.isRequired,
  source: PropTypes.string.isRequired,
};
/*
const mapStateToProps = state => ({});

const mapDispatchToProps = (dispatch: Function, props): Object => ({
  goToTaskDetails: id => {
    dispatch(pushRoute({
      key: 'taskDetails',
      id,
    }));
  },
});

const VenueConnected = connect(
  mapStateToProps,
  mapDispatchToProps,
)(VenueDetails);
*/

const VenueDetailsQuery = gql`
  query VenueDetails($id: ID!, $source: String) {
    venue(id: $id, source: $source) {
      ...VenueDescription
      ...Tasks
    }
  }
  ${VenueDescription.fragments.venue}
  ${Tasks.fragments.venue}
`;


const AddTaskMutation = gql`
  mutation task($taskId: ID!, $task: TaskInput!, $token: String!) {
    task(venueId: $taskId, task: $task, token: $token) {
      title
      nbAnswers
    }
  }
`;

function mapStateToProps(state) {
  return {
    ...state.user,
  };
}

const VenueDetailsConnected = connect(mapStateToProps)(VenueDetails);

export default graphql(AddTaskMutation, {
  props: ({ownProps, mutate}) => ({
      addTask: (taskId, task, token) => mutate({
        variables: {taskId, task, token},
        optimisticResponse: {
          __typename: 'Mutation',
          task: {
            __typename: 'Task',
            task: task,
            date: new Date(),
            postedBy: {
              username: '',
            },
          },
        },
        updateQueries: {
          VenueDetails: (prev, { mutationResult }) => {
            const newTask = mutationResult.data.task;
            const existing = null; // prev.task.answers.find(existingAnswer => existingAnswer.answer === newAnswer.answer);
            /*
            if (existing) {
              return prev;
            }
            */
            return update(prev, {
              venue: {
                tasks: { $push: [newTask] },
              },
            });
          },
        },
      }),
  }),
})(graphql(VenueDetailsQuery, {
  options: ({ _id, sourceId, source }) => {
    const variables = {};
    if (_id) {
      variables.id = _id;
      variables.source = null;
    } else {
      variables.id = sourceId;
      variables.id = source;
    }
    return {variables};
  },
  props: ({ownProps, data: { loading, error, venue } }) => {
    return {
      isFetching: loading,
      venue: {
        _id: ownProps._id,
       ...venue,
      },
      error,
      navigator: ownProps.navigator,
    };
  },
})(VenueDetailsConnected));
