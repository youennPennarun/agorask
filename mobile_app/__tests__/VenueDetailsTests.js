/* global jest, expect, describe, it */

import 'react-native';
import React from 'react';

import {VenueDetails} from '../src/components/views/venueDetails/VenueDetails';

import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

const venueTest = {
  _id: '5834b73cb5cb2c65d95f7cea',
  name: 'Test venue',
  address: {
    formatted: [
      '2 univ street',
      '5200 Belfast',
      'UK',
    ],
  },
  categories: [],
  contact: {
    formattedPhone: '(060) 25-65',
  },
};

it('renders correctly', () => {
  const wrapper = shallow(
    <VenueDetails sourceId='00000'
      source='source'
      getVenueDetails={() => { console.log('getVenueDetails'); }}
      venue={venueTest} />
  );
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});
