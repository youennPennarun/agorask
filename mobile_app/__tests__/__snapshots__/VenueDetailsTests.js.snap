exports[`test renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#e9e9e9",
      "flex": 1,
    }
  }>
  <Image
    resizeMode="cover"
    source={
      Object {
        "uri": "http://192.168.0.10:3000/venues/5834b73cb5cb2c65d95f7cea/image",
      }
    }
    style={
      Object {
        "height": 200.1,
        "width": 375,
      }
    } />
  <ScrollView
    style={Object {}}>
    <VenueDescription
      venue={
        Object {
          "_id": "5834b73cb5cb2c65d95f7cea",
          "address": Object {
            "formatted": Array [
              "2 univ street",
              "5200 Belfast",
              "UK",
            ],
          },
          "categories": Array [],
          "contact": Object {
            "formattedPhone": "(060) 25-65",
          },
          "name": "Test venue",
        }
      } />
    <Tasks
      goToTask={[Function]}
      tasks={Array []} />
  </ScrollView>
</View>
`;
